### ISSUE TITLE
want to retrain a model with new images

### ISSUE THREAD

**Issue by sourabCV** [ultralytics/ultralytics#22390]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I want to retrain my old model with new captured images.
4 months ago, i trained a model with yolov12 with classes {A, B, C}
now a days i found my model is not predicting good for class A, so i recollected some images only for class A.
I dont have my old dataset so i want to improve my model best.pt by retraining with new images.

i tried 
```
from ultralytics import YOLO
model = YOLO('/content/drive/MyDrive/a/epoch50.pt')
results = model.train(data='/content/drive/MyDrive/a/dfyolo.yaml', epochs=20, imgsz=640,batch=2,device=[0],lr0=0.001,close_mosaic = 0,pretrained="/content/drive/MyDrive/a/epoch50.pt")  
```

but after this retraining its just detecting class A not showing class B, C why ?
 suggest me how to retrain ?

### Additional

1 try 
```
from ultralytics import YOLO
model = YOLO('/content/drive/MyDrive/a/epoch50.pt')
results = model.train(data='/content/drive/MyDrive/a/dfyolo.yaml', epochs=20, imgsz=640,batch=2,device=[0],close_mosaic = 0,pretrained="/content/drive/MyDrive/a/epoch50.pt")  
```
2 try 
```
from ultralytics import YOLO
model = YOLO('/content/drive/MyDrive/a/epoch50.pt')
results = model.train(data='/content/drive/MyDrive/a/dfyolo.yaml', epochs=20, imgsz=640,batch=2,device=[0],lr0=0.001,close_mosaic = 0,pretrained="/content/drive/MyDrive/a/epoch50.pt")  
```

both result not storing old dataset detection and i dont have my old dataset to merge and retrain
pls help

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sourabCV, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will follow up soon to assist you further.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (e.g., retraining/fine-tuning on new data), please include as much detail as possible so we can help:
-Exact `ultralytics` package version and environment (OS, Python, PyTorch, CUDA)
-The exact training command/code used and full logs
-Your `data.yaml` (class names and splits) and a few sample images per class
-Which checkpoint you started from and where your results/weights are saved
-Confirmation you reviewed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) ✅

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us guide you on retraining with new images and preserving performance across classes, please share the requested details above along with the exact `data.yaml` and a few representative examples for classes A, B, and C 📦🧪

**Comment by Y-T-G** [comment]
You will need the whole dataset for retraining otherwise it will forget the old classes. It's just how gradient descent works.

You can try linear probing by adding this code before you start training:

```
# Identify the head layer index
head_index = len(model.model.model) - 1

# Freeze all backbone and neck layers (i.e. everything before the head)
freeze = [str(i) for i in range(0, head_index)]

# Freeze parts of the segmentation head, keeping only the classification branch trainable
for name, child in model.model.model[-1].named_children():
    if "cv3" not in name:
        freeze.append(f"{head_index}.{name}")

# Freeze detection branch components
freeze.extend(
    [
        f"{head_index}.cv3.0.0",
        f"{head_index}.cv3.0.1",
        f"{head_index}.cv3.1.0",
        f"{head_index}.cv3.1.1",
        f"{head_index}.cv3.2.0",
        f"{head_index}.cv3.2.1",
    ]
)
```
and then passing `freeze=freeze` to `model.train()`

You can also try training a model with an extra head and everything else frozen. And average the predictions for the trained class from old and new head. It will make the speed slower, but it will not forget the old classes. You will need to modify the Ultralytics source code for this to work.

**Comment by sourabCV** [comment]
```
from ultralytics import YOLO
model = YOLO('/content/drive/MyDrive/msil_hoodline_131025/epoch50.pt')
# Identify the head layer index
head_index = len(model.model.model) - 1

# Freeze all backbone and neck layers (i.e. everything before the head)
freeze = [str(i) for i in range(0, head_index)]

# Freeze parts of the segmentation head, keeping only the classification branch trainable
for name, child in model.model.model[-1].named_children():
    if "cv3" not in name:
        freeze.append(f"{head_index}.{name}")

# Freeze detection branch components
freeze.extend(
    [
        f"{head_index}.cv3.0.0",
        f"{head_index}.cv3.0.1",
        f"{head_index}.cv3.1.0",
        f"{head_index}.cv3.1.1",
        f"{head_index}.cv3.2.0",
        f"{head_index}.cv3.2.1",
    ]
)
results = model.train(data='/content/drive/MyDrive/msil_hoodline_131025/dfyolo.yaml', epochs=20, imgsz=640,batch=2,device=[0],freeze=freeze,close_mosaic = 0,pretrained="/content/drive/MyDrive/msil_hoodline_131025/epoch50.pt")
```


training started with this, will update after result

**Comment by Y-T-G** [comment]
Your dataset YAML should contain all the classes, not just one class. Otherwise, the head will be reset. It should contain all the classes in the same order and class ID like in your original dataset YAML.

**Comment by sourabCV** [comment]
yes i mention all classes names in YAML file

**Comment by sourabCV** [comment]
<img width="1520" height="918" alt="Image" src="https://github.com/user-attachments/assets/6dce747b-6d55-431b-9a80-ebee5a58adf9" />
after 50 epoch also model is not improving with new images 
its showing same output and missing some objects like old model

**Comment by Y-T-G** [comment]
You can reduce the LR and set optimizer to AdamW. The batch size shouldn't be so low.

**Comment by sourabCV** [comment]
i have 2454 train images  1051 val images 

based on that pls suggest me what changes needed in model.train 

what values going to be perfect

**Comment by glenn-jocher** [comment]
With 2.5k/1k images, try a simple fine-tune baseline (don’t disable mosaic) and let the trainer auto-size the batch; example:

```python
from ultralytics import YOLO

model = YOLO('/content/drive/MyDrive/msil_hoodline_131025/epoch50.pt')

results = model.train(
    data='/content/drive/MyDrive/msil_hoodline_131025/dfyolo.yaml',
    epochs=80,
    imgsz=640,
    batch=-1,              # auto-tune batch
    optimizer='AdamW',
    lr0=1e-3,              # Adam/AdamW-friendly LR
    weight_decay=5e-4,
    patience=30,
    cos_lr=True,
    close_mosaic=10,       # keep default schedule
    device=0
)
```

If you only add class-A images the model will drift; add a small replay set for classes B and C (even 50–100 images/class) to maintain them. Training knobs are detailed in the Train guide and settings table: see the [Model Training page](https://docs.ultralytics.com/modes/train/) and the [Configuration reference](https://docs.ultralytics.com/usage/cfg/). If possible, consider re-finetuning with Ultralytics YOLO11, our recommended stable baseline, using the same recipe.

**Comment by DoubleIsTrouble** [comment]
@sourabCV I found an [issue](https://github.com/ultralytics/ultralytics/issues/5135) that’s somewhat similar to your situation — it might be helpful to you. #5135
